src/ImageSharp/Formats/Qoi/QoiEncoderCore.cs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) Six Labors.
+// Licensed under the Six Labors Split License.
+
+using System.Buffers.Binary;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+
+namespace SixLabors.ImageSharp.Formats.Qoi;
+
+/// <summary>
+/// Image encoder for writing an image to a stream as a QOi image
+/// </summary>
+public class QoiEncoderCore : IImageEncoderInternals
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="QoiEncoderCore"/> class.
+    /// </summary>
+    public QoiEncoderCore()
+    {
+    }
+
+    /// <inheritdoc />
+    public void Encode<TPixel>(Image<TPixel> image, Stream stream, CancellationToken cancellationToken)
+        where TPixel : unmanaged, IPixel<TPixel>
+    {
+        Guard.NotNull(image, nameof(image));
+        Guard.NotNull(stream, nameof(stream));
+
+        WriteHeader(image, stream);
+        WritePixels(image, stream);
+        WriteEndOfStream(stream);
+        stream.Flush();
+    }
+
+    private static void WriteHeader(Image image, Stream stream)
+    {
+        // Get metadata
+        Span<byte> width = stackalloc byte[4];
+        Span<byte> height = stackalloc byte[4];
+        BinaryPrimitives.WriteUInt32BigEndian(width, (uint)image.Width);
+        BinaryPrimitives.WriteUInt32BigEndian(height, (uint)image.Height);
+        QoiChannels qoiChannels = image.PixelType.BitsPerPixel == 24 ? QoiChannels.Rgb : QoiChannels.Rgba;
+
+        // I need to check this, how do I check it with the pixel type or metadata of the original image?
+        const QoiColorSpace qoiColorSpace = QoiColorSpace.SrgbWithLinearAlpha;
```

Comment:
We probably want to make this configurable by properties exposed on `QoiEncoder`. Check `PngEncoder` as an example,  eg:

https://github.com/SixLabors/ImageSharp/blob/2738b3ab5793bddb96f3a5317d46afa0529220ce/src/ImageSharp/Formats/Png/PngEncoderCore.cs#L1257

Reply:
I've done the changes, however, now `WriteHeader(Image image, Stream stream)` method can't be static but `WritePixels` and `WriteEndOfStream` are, isn't there any problem with that?

Reply:
No, that's an implementation detail and it's fine.